### ISSUE TITLE
Pipelining & Notify/Awake with a message

### ISSUE THREAD

**Issue by arthurprs** [tailhook/rotor-stream#4]
Is is possible to implement some sort of pipeline in protocols built on top of rotor-stream? My understanding is that it isn't, since there's no intent for both receiving/sending. Is that correct?

**Comment by tailhook** [comment]
It's definitely possible. It's not possible currently in rotor-http, because we tried to make as little memory allocations as possible (i.e. to have pipelining, you need to have a list of requests and so on).

Actually, I'm not sure what could be the problem with pipelining with rotor-stream. You just have to never return  `Expectation::Sleep` or `Expectation::Flush` (you don't need to "expect" flush for writing, only if you want TCP pushback, writing happens asynchronously after you write to the output buffer). And always return expectation for bytes.

If you explain your use case better, I could help you with the issues if there are any.

**Comment by arthurprs** [comment]
Thank you! I didn't know I didn't have to return flush() in order to mark the socket as "sendable" in mio. I did some tests, the library is great!

I miss notifying the Loop/State Machine with a message directly. Is this something you plan to add in the future? Right now I have to notify it and then on wakeup pop some queue, that is wasteful since I basically need a queue per State Machine.

**Comment by tailhook** [comment]
> I miss notifying the Loop/State Machine with a message directly. Is this something you plan to add in the future? Right now I have to notify it and then on wakeup pop some queue, that is wasteful since I basically need a queue per State Machine.

Are you talking about [handler.notify()](http://rustdoc.s3-website-us-east-1.amazonaws.com/mio/v0.5.x/mio/trait.Handler.html#method.notify) ?

The problem to make it is that it doesn't compose easily. I.e. if you have a http state machine and a redis state machine. You want to put them in the same loop. You have to compose state machine **and** message. And this is just a bigger boilerplate not a real issue. Real issue follows.

Let's pretend we have these enums:

``` rust
enum Fsm { Http(HttpFsm), Redis(RedisFsm) }
enum Message { Http(HttpMsg), Redis(RedisFsm) }
```

This mean that to send `HttpMsg::Reply` from `RedisFsm`, the `RedisFsm` you need to send `Message::Http(http_reply)`. Which means your `RedisFsm` implementation must know about `Message`.

This is solvable by `Box`-ing. But I think it's more wasteful than using a queue per state machine.

Also, you don't have to have a queue per state machine. You need a queue per message type. I.e. if you have a pull of state machines with workers, you may have one queue per pool. To make this work, you need a queue in context (or `Arc<Queue>` somewhere) and a collection of `Notifier` of state machines to wake up. 

If you have any better ideas of how solve messaging, I open to suggestions.

**Comment by arthurprs** [comment]
I understand the reasoning and I agree it doesn't always work. But when it does it's very useful. I'm probably missing something but it's very usable in most cases I can think of.

```
enum Fsm { Http(HttpFsm), Redis(RedisFsm) }
enum Message { Http(HttpMsg), Redis(RedisMsg) }
```

You could have Message in another module and make everyone use it. This is what most mio users do anyway.

One may argue that keeping the status quo prevents the user from having problems like the one you're describing, other might say it does more good than harm. Tricky.

**Comment by arthurprs** [comment]
As for the queue per message type. I don't think I understand your solution.

In the context of this https://github.com/brson/basic-http-server/ for example, multiple file-io threads may race between putting messages into the queue and awaking the state machine that should pop it.

**Comment by tailhook** [comment]
> You could have Message in another module and make everyone use it. This is what most mio users do anyway.

Let me try to elaborate on that. Let's pretend we have `http_proxy` crate which have a message type:

``` rust
enum Message {
  Request(Req),   // Sent from `http_proxy::ServerFsm` to `http_proxy::ClientFsm`
  Response(Resp). // Sent back from client to server fsms
}
```

And you embed the proxy into application, effectively having:

``` rust
enum Message {
  Proxy(http_proxy::Message),
  Redis(MyRedisMessage),
}
```

Now, the code in `http_proxy::ServerFsm` must send `your_app::Message` and not `http_proxy::Message`. How would you achieve that?

Is the issue is more clear now?

> In the context of this https://github.com/brson/basic-http-server/ for example, multiple file-io threads may race between putting messages into the queue and awaking the state machine that should pop it.

Yes. But if you have two tasks queued and two guaranteed wake ups, you don't care which state machines will pick up which tasks in which order. I mean it works only in cases where it doesn't matter.

In the context of an HTTP proxy server, it would work for messages from server socket, to a client connection pool. But not for responses.

For responses, I believe you have to have a `Vec` for each client connection anyway, because:
1. Your responses may arrive out of order (or why would you pipeline them anyway), so you have check which request this response is for. And you have to queue out of order responses.
2. I think you need a vec to handle timeouts properly. I.e. to send 502 if you can't respond in time (this is just at the top of my head, maybe there is a simpler solution).

And once you have a Vec of request/responses in a state machine you may just use it as a per-state machine queue.

**Comment by arthurprs** [comment]
Thanks, I see what you mean. Unfortunately I can't think of a solution neither :disappointed:

**Comment by tikue** [comment]
I think rotor-stream makes the wrong call here. Message notifications is important for even quite basic async applications. A typical use case is to farm out the business logic to a thread pool and have the main event loop doing purely IO. But to do that, you need to be able to send responses back to the event loop when they're ready to be written. You can't do that without notify. Dropbox engineers working on Magic Pocket have [stated](https://www.reddit.com/r/rust/comments/4l2hlr/async_io_for_rust/d3k7fju?context=3) that they use MIO in this fashion.

I understand that you want machines to compose easily, but I'm reminded of the adage "Easy things should be easy, and hard things should be possible."

**Comment by tikue** [comment]
I think the right way to do this is to allow the impler to specify the message type. If different libraries want to be maximally composable, they can just specify the message type to be `()`.

**Comment by tailhook** [comment]
@tikue, sure, I understand your argumentation. I've tried several options during first versions of rotor. 

I think you can perceive the current abstraction similarly to a [futex](http://man7.org/linux/man-pages/man2/futex.2.html) or a [parking_lot](https://amanieu.github.io/parking_lot/parking_lot/fn.park.html). I.e. the super low-level abstraction. We can build the queues and other abstractions on top.

I think you want global queue just because you had come from the mio world where queue is built-in (well, I'm not sure if it's event a part of the loop in the current mio master or just build on top). But in other libraries there is no such thing, at the top of my head: in `libev` the closest thing is `ev_async` which is very similar to `rotor::Notifier` we have here, in the python's `asyncio` there are futures (but not a way to send an arbitrary message to arbitrary coroutine), so on.

> But to do that, you need to be able to send responses back to the event loop when they're ready to be written. You can't do that without notify

Sure. There is a `Scope::notifier()` which makes a wake-up-this-state-machine object. You can make a future or a queue using this primitive. There is a [future primitive in rotor-tools](http://tailhook.github.io/rotor-tools/rotor_tools/future/index.html), but I'm not sure it's a good implementation, so it's not documented.

We should probably make some higher-level things like queues and futures to make working with it easier, but there are a lot of questions to each of the abstraction, so I have not come up with good generic ones. Anyway, as you can see generic future I've linked above is about hundred lines of code. And you can probably make non-generic one for your app much shorter.

Hopefully, this makes the situation a little bit clearer.

**Comment by tikue** [comment]
But rotor is built on top of mio, so it's not actually a lower level of abstraction. There's still a `Queue<()>` underneath, but now there's also another `Queue<Message>` that every user has to make themselves. And now there are two queue polls, not one, for each notify.

Just to reiterate what I said in my second comment: if a library built on rotor wants to be composable with other libraries, they can always set `type Message = ()` and go the global queue route as you've described. So it's actually maximally flexible to allow the user to specify message type.

**Comment by tailhook** [comment]
> There's still a Queue<()> underneath,

There is. But it's an implementation detail. I'm seeking for the ways to get rid of it in favor of something more efficient (i.e. we need a bit for each state machine not a word or two)

> but now there's also another Queue<Message> that every user has to make themselves. And now there are two queue polls, not one, for each notify.

So check of the queue is just a read of the pointer and jump. It isn't adding too much when we need to read a state of a state machine do something with a message and so on. I doubt it has any noticeable performance difference.

Moreover, we optimize high throughput scenarios by only using Notifier once. So writer puts a message just into one queue if that queue already has elements (i.e. we already had notified state machine, but it has not woken up yet).

In many scenarios like you've described where you put a reply to a request from the pull of workers, you don't even need a queue, just a future. Or even better: you can put response directly into the response buffer and wakeup a state machine.

Maybe you are misunderstanding what that queue is?

---

On the other hand, when you use notification queue in mio how do you size it? I.e. in my opinion the fixed-size mio queue is just ticking bomb: i.e. 90% of the time for 1000 state machines you only have 100 messages queued at the time. But on the peak load, you may need 2000 so you have to always reserve 2000 slots. But it's still not a guarantee. You may receive more and get a crash.

I mean even if throw away composition, I don't see a reasonable way to use the notification queue as queue.

**Comment by tikue** [comment]
Yeah, the fixed size aspect has been a pain point for me, as well. But isn't MIO changing to allow using custom queues?

**Comment by tailhook** [comment]
>  But isn't MIO changing to allow using custom queues?

I'm not sure if I fully understand recent changes in mio. It's likely yes. And if it's true, why don't we allow each library to register it's own queue instead of making one global?
